File: web/src/types/invoice.ts
export type InvoiceStatus = 'processing' | 'pending' | 'failed' | 'confirmed' | 'archived';

export type InvoiceType = '增值税' | '高铁' | '滴滴' | '飞机' | '其他';

export interface Invoice {
  id: number;
  user_id: number;
  invoice_no: string | null;
  amount: number | null;
  invoice_date: string | null;
  category: string | null;
  vendor: string | null;
  buyer_name: string | null;
  invoice_type: string | null;
  project_name: string | null;
  train_no: string | null;
  departure_station: string | null;
  arrival_station: string | null;
  departure_location: string | null;
  arrival_location: string | null;
  flight_no: string | null;
  departure_city: string | null;
  arrival_city: string | null;
  file_path: string;
  file_original_name: string | null;
  status: InvoiceStatus;
  remark: string | null;
  ocr_raw_data: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
}

export interface UpdateInvoiceRequest {
  invoice_no?: string | null;
  amount?: number | null;
  invoice_date?: string | null;
  category?: string | null;
  vendor?: string | null;
  buyer_name?: string | null;
  invoice_type?: string | null;
  project_name?: string | null;
  train_no?: string | null;
  departure_station?: string | null;
  arrival_station?: string | null;
  departure_location?: string | null;
  arrival_location?: string | null;
  flight_no?: string | null;
  departure_city?: string | null;
  arrival_city?: string | null;
  remark?: string | null;
}

export interface UploadFileResult {
  filename: string;
  success: boolean;
  invoice_id: number | null;
  error: string | null;
}

export interface UploadResponse {
  results: UploadFileResult[];
  skipped_count: number;
}

export interface InvoiceListResponse {
  items: Invoice[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface DeleteResponse {
  deleted: boolean;
  type: 'hard' | 'soft';
  restorable_until: string | null;
}

export interface InvoiceListParams {
  state?: string;
  page?: number;
  page_size?: number;
  date_from?: string;
  date_to?: string;
}

File: server/app/schemas/invoice.py
from datetime import date, datetime

from pydantic import BaseModel, Field


class InvoiceResponse(BaseModel):
    id: int
    user_id: int
    invoice_no: str | None
    amount: float | None
    invoice_date: date | None
    category: str | None
    vendor: str | None
    file_path: str
    file_original_name: str | None
    status: str
    buyer_name: str | None
    invoice_type: str | None
    project_name: str | None
    ocr_raw_data: dict | None
    remark: str | None
    train_no: str | None
    departure_station: str | None
    arrival_station: str | None
    departure_location: str | None
    arrival_location: str | None
    flight_no: str | None
    departure_city: str | None
    arrival_city: str | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class UpdateInvoiceRequest(BaseModel):
    invoice_no: str | None = Field(default=None, max_length=50)
    amount: float | None = None
    invoice_date: date | None = None
    category: str | None = Field(default=None, max_length=50)
    vendor: str | None = Field(default=None, max_length=200)
    remark: str | None = Field(default=None, max_length=500)
    buyer_name: str | None = Field(default=None, max_length=200)
    invoice_type: str | None = Field(default=None, max_length=100)
    project_name: str | None = Field(default=None, max_length=200)
    train_no: str | None = Field(default=None, max_length=50)
    departure_station: str | None = Field(default=None, max_length=100)
    arrival_station: str | None = Field(default=None, max_length=100)
    departure_location: str | None = Field(default=None, max_length=200)
    arrival_location: str | None = Field(default=None, max_length=200)
    flight_no: str | None = Field(default=None, max_length=50)
    departure_city: str | None = Field(default=None, max_length=100)
    arrival_city: str | None = Field(default=None, max_length=100)


class UploadFileResult(BaseModel):
    filename: str
    success: bool
    invoice_id: int | None = None
    error: str | None = None


class UploadResponse(BaseModel):
    results: list[UploadFileResult]
    skipped_count: int = 0


class InvoiceListResponse(BaseModel):
    items: list[InvoiceResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class DeleteResponse(BaseModel):
    deleted: bool
    type: str
    restorable_until: datetime | None = None
